Model/Api/OAuthRedirect.php

```diff
@@ -35,37 +46,198 @@ class OAuthRedirect implements OAuthRedirectInterface
+        try {
+            $externalCustomerEntity = $this->externalCustomerEntityRepository->getByExternalID($payload['sub']);
+        } catch (NoSuchEntityException $nsee) {
+        }
+        try {
+            if ($externalCustomerEntity !== null) {
+                $customer = $this->customerRepository->getById($externalCustomerEntity->getCustomerID());
+            } else {
+                $customer = $this->customerRepository->get($payload['email'], $websiteId);
+            }
+        } catch (NoSuchEntityException $nsee) {
+        }
+
+        // External customer entity exists, but the customer it links to doesn't
+        // This should happen very rarely, if at all, which is why we notify bugsnag when this happens
+        if ($externalCustomerEntity !== null && $customer === null) {
+            $this->bugsnag->notifyError(
+                'OAuthRedirect',
+                'external customer entity ' . $payload['sub'] . ' linked to nonexistent customer ' . $externalCustomerEntity->getCustomerID()
+            );
+        }
+
+        // If the customer isn't linked and it exists in M2, but the email is not verified, we throw an exception
+        if ($externalCustomerEntity === null && $customer !== null && !$payload['email_verified']) {
+            throw new WebapiException(__('Internal Server Error'), 0, WebapiException::HTTP_INTERNAL_ERROR);
+        }
+
+        try {
+            $customer = $customer ?: $this->createNewCustomer($websiteId, $storeId, $payload);
+            $this->linkAndLogin($payload['sub'], $customer->getId());
+        } catch (Exception $e) {
+            throw new WebapiException(__('Internal Server Error'), 0, WebapiException::HTTP_INTERNAL_ERROR);
+        }
+    }
+
+    /**
+     * Create new customer from provided parameters
+     *
+     * @param int   $websiteId
+     * @param int   $storeId
+     * @param mixed $payload
+     */
+    private function createNewCustomer($websiteId, $storeId, $payload)
+    {
+        $newCustomer = $this->customerInterfaceFactory->create();
+        $newCustomer->setWebsiteId($websiteId);
+        $newCustomer->setStoreId($storeId);
+        $newCustomer->setFirstname($payload['first_name']);
+        $newCustomer->setLastname($payload['last_name']);
+        $newCustomer->setEmail($payload['email']);
+        $newCustomer->setConfirmation(null);
+        return $this->customerRepository->save($newCustomer);
+    }
+
+    /**
+     * Link the external ID to customer if needed, log in, and redirect
+     *
+     * @param string $externalID
+     * @param int    $customerID
+     */
+    private function linkAndLogin($externalID, $customerID)
+    {
+        $this->externalCustomerEntityRepository->upsert($externalID, $customerID);
+        $customerModel = $this->customerFactory->create()->load($customerID);
+        $this->customerSession->setCustomerAsLoggedIn($customerModel);
+        $this->response->setRedirect($this->url->getAccountUrl())->sendResponse();
```

Comment:
maybe move this line out, it doesn't look like `linkAndLogin`

Reply:
well i'll rename the method name then